nix/postgres.nix

```diff
@@ -236,153 +231,159 @@ in
+                ''
+                  export PGDATA="${config.dataDir}"
+                  ${lib.getExe setupScript}
+                '';
 
-        # DB process
-        ${cfg.name} =
-          let
-            startScript = pkgs.writeShellApplication {
-              name = "start-postgres";
-              text = ''
-                set -x
-                export PATH="${postgresPkg}"/bin:$PATH
-                PGDATA=$(readlink -f "${cfg.dataDir}")
-                export PGDATA
-                postgres -k "$PGDATA"
-              '';
+              # DB process
+              ${name} =
+                let
+                  startScript = pkgs.writeShellApplication {
+                    name = "start-postgres";
+                    text = ''
+                      set -x
+                      export PATH="${postgresPkg}"/bin:$PATH
+                      PGDATA=$(readlink -f "${config.dataDir}")
+                      export PGDATA
+                      postgres -k "$PGDATA"
+                    '';
+                  };
+                in
+                {
+                  command = startScript;
+                  depends_on."${name}-init".condition = "process_completed_successfully";
+                  # SIGINT (= 2) for faster shutdown: https://www.postgresql.org/docs/current/server-shutdown.html
+                  shutdown.signal = 2;
+                  readiness_probe = {
+                    exec.command = "${postgresPkg}/bin/pg_isready -h $(readlink -f ${config.dataDir}) -d template1";
+                    initial_delay_seconds = 2;
+                    period_seconds = 10;
+                    timeout_seconds = 4;
+                    success_threshold = 1;
+                    failure_threshold = 5;
+                  };
+                  # https://github.com/F1bonacc1/process-compose#-auto-restart-if-not-healthy
+                  availability.restart = "on_failure";
+                };
             };
-          in
-          {
-            command = startScript;
-            depends_on."${cfg.name}-init".condition = "process_completed_successfully";
-            # SIGINT (= 2) for faster shutdown: https://www.postgresql.org/docs/current/server-shutdown.html
-            shutdown.signal = 2;
-            readiness_probe = {
-              exec.command = "${postgresPkg}/bin/pg_isready -h $(readlink -f ${cfg.dataDir}) -d template1";
-              initial_delay_seconds = 2;
-              period_seconds = 10;
-              timeout_seconds = 4;
-              success_threshold = 1;
-              failure_threshold = 5;
-            };
-            # https://github.com/F1bonacc1/process-compose#-auto-restart-if-not-healthy
-            availability.restart = "on_failure";
-          };
+        };
       };
+    }));
+  };
+  config = let mergeMapAttrs = f: attrs: lib.mkMerge (lib.mapAttrsToList f attrs); in {
+    settings.processes = mergeMapAttrs (name: cfg: lib.mkIf cfg.enable cfg.outputs) config.services.postgres;
```

Comment:
Can't this be written using `imports`?

```nix
{
  settings.processes = {
    imports = lib.mapAttrs (_: cfg: lib.mkIf cfg.enable cfg.outputs) config.service.postgres;
  };
}
```

Reply:
I get an error complaining about `settings.processes.imports` not being of type `submodule`

Reply:
We should probably output the entire settings then (`outputs.settings`); making it an import-eable module provides the advantage of letting the module system deal with conflicts.

Reply:
You mean then we can do something like:
```nix
{
  config.settings.imports = lib.mapAttrsToList (_: cfg: lib.mkIf cfg.enable cfg.outputs) config.services.postgres;
}
```
Correct me if I am wrong, that shouldn't be possible because `imports` option doesn't exist in `process-compose.default.settings`.

Reply:
My error. I meant the top-level submodule. So, `output` option would set the _whole_ process-compose config (with the expectation that they are normally contain _only_ `processes`).

The you can do,

{
  config = {
    imports = ...;
  };
}
```

Reply:
Though I wonder if we should create `settings` using `submoduleWith` (to avoid [`shorthandOnlyDefinesConfig`](https://github.com/NixOS/nixpkgs/blob/78386f6b895bb3afb33732dbf8d0eb6dae809194/lib/types.nix#L721-L725) disabling imports) as well ...

Reply:
Defining either `settings` or `processes` option with `submoduleWith` looks like the right direction, I will try this out.